### ISSUE TITLE
Promise detection does not work in electron-like framework environments.

### ISSUE THREAD

**Issue by YuriSolovyov** [mozilla/localForage#376]
Error happens here:
```javascript
    // Promises!
    var Promise = (typeof module !== 'undefined' && module.exports) ?
                  require('promise') : this.Promise;
    //             ^----- errors here
```
Error happens because `electron` and `nw.js` window contexts unlike browsers have their own `module`, `module.exports` and `require` objects.

In this case, `require` works like node-style require, not requireJS one, so it tried to find `'promise'` module in node_modules folder, and of course there is not such module there.

So if you add script to `.html` page with
```html
<script defer type="text/javascript" src="path/to/localforage.min.js"></script>
```
it will fail:
```
Uncaught Error: Cannot find module 'promise'
Module._resolveFilename @
module.js:324Module._load @
module.js:266Module.require @
module.js:353require @
module.js:372(anonymous function) @
localforage.min.js:246(anonymous function) @
localforage.min.js:625
```
Tried both `localforage.min.js` and `localforage.nopromises.min.js`.

electron: "0.25.3"

**Comment by lordgreg** [comment]
This is exact same problem we are having when building Electron app. I've read about it in my 2nd comment here: https://discuss.atom.io/t/electron-build-not-finding-js-functions-pack/20549/2

Is this going to be fixed/solved any time soon?

**Comment by thgreasi** [comment]
This kind of detection and requiring Promises has been removed in the current master.
Any chance you can try using the code in the current master branch and report back?

Requiring polyfills is considered an anti-pattern by lost of devs. A library just should not dictate the use of a polyfill (like we did in v1.2) but rather expect that it is already loaded if required.
As a result, we provide both a dist file w/ and w/o the polyfill, but it's up to the integrator to ensure that the environment is proper for the lib to work.

**Comment by lordgreg** [comment]
I will report the status as soon as I test it. Until then, thank you! :+1:

**Comment by lordgreg** [comment]
I've installed latest localforage from master branch like this:

`bower install localforage#master`.

Before that, the errors I've recieved were:

```
Uncaught Error: Cannot find module 'promise' (module.js:338) and
Uncaught Error: Cannot find module 'localForage' (module.js:338)
```

After updating localforage from master, only one error appears:

```
Uncaught Error: Cannot find module 'localForage' (module.js:338)
```

...but the promise error is gone.

**Comment by tofumatt** [comment]
We no longer require promises lib as mentioned in master, but you'll need to use the dist version of localforage now. Which file are you trying to use?



- tofumatt

**Comment by lordgreg** [comment]
`<script src="bower_components/localforage/dist/localforage.js"></script>` which has this header:

```
/*!
    localForage -- Offline Storage, Improved
    Version 1.2.10
    https://mozilla.github.io/localForage
    (c) 2013-2015 Mozilla, Apache License 2.0
*/
```

Should I take any other?

**Comment by tofumatt** [comment]
Nope, that should be the right file.




What browser are you using? This looks like a bug for sure! (For what it’s worth, I would say stick with 1.2, which is our official release right now). master is tested but right now we’re upgrading some of our distribution methods so things might be unusually unstable! ;-)


- tofumatt

**Comment by lordgreg** [comment]
As told by OP, this is Electron packager (http://electron.atom.io/), and as their Github page states (https://github.com/atom/electron) `It is based on io.js and Chromium`

**Comment by tofumatt** [comment]
Oh, right. I’m sorry, I totally missed that.

We don’t aim to have node.js (io.js) support for localForage (https://github.com/mozilla/localForage/blob/master/CONTRIBUTING.md#nodejs-support). It’s aimed to be a browser library first and foremost. I can’t really offer much support for node usage.

If it works that’s great but if it doesn’t we want to focus on localForage issues that affect client browsers and things like PhoneGap. Because Electron is essentially using a node “backend” localForage isn’t supported there (and I’m not even sure would work, given there aren’t any storage methods available to node like IndexedDB).

**Comment by tofumatt** [comment]
So unless you're using it in the Chromium part of the app I don't think this will work. Maybe you can clarify for me. Thanks!

**Comment by tofumatt** [comment]
Oh, right. I’m sorry, I totally missed that.




We don’t aim to have node.js (io.js) support for localForage (https://github.com/mozilla/localForage/blob/master/CONTRIBUTING.md#nodejs-support). It’s aimed to be a browser library first and foremost. I can’t really offer much support for node usage.




If it works that’s great but if it doesn’t we want to focus on localForage issues that affect client browsers and things like PhoneGap. Because Electron is essentially using a node “backend” localForage isn’t supported there (and I’m not even sure would work, given there aren’t any storage methods available to node like IndexedDB).


- tofumatt

**Comment by YuriSolovyov** [comment]
@tofumatt it is not about node.js support, the problem will occur even without using any node-specific things. Even simpliest examples are not working. Electron does not use node for anything like db backend or something.

It is just about how you detect require.js, problem is that `require` in localforage collides with `require` in Electron, that's it.

**Comment by thgreasi** [comment]
@lordgreg Can you please provide the line in localforage that causes the error when using the current master branch?

**Comment by lordgreg** [comment]
Hi. Here are my results for the past day:

1. As you've suggested, I've updated localforage from master branch (since version 1.2.10 didn't worked) and the `promise not found` was gone. The promise problem was also causing out app not to work at all.

2. Then we still had to figure out, where does the `localforage not found` problem comes. The problem is, there was no error form localforage.js but direct from `module.js`. We found out, that when we were using `angular-localForage`. Even after updating localforage.js (to 1.2.4, 1.2.10 or master), we've always received above error.

We've completely removed angular-localForage from project and updated localforage.js to latest commit (b91dfa532b2a62ff344188e165b4878b0efe7574). Now, we are using:

`return $q.when($window.localforage.setItem('xyz', this.xyz)`

and

`return $q.when($window.localforage.getItem('xyz')`

instead of angular promises. we're using $q and localforage directly. Now, everything works. 

:+1: 

Thank you Mozilla devs :)
